domain/permission/state/state.go

```diff
@@ -188,59 +189,187 @@ WHERE grant_to = $M.grant_to AND grant_on = $M.grant_on
+	var userAccess corepermission.UserAccess
+	db, err := st.DB()
+	if err != nil {
+		return userAccess, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err := st.findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccess = user.toCoreUserAccess()
+
+		// Based on the grant to and grant from, find the permission,
+		// then the access type of it.
+		accessType, permissionUUID, err := st.findAccessType(ctx, tx, target.Key, user.UUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccess.Access = corepermission.Access(accessType)
+		userAccess.PermissionID = permissionUUID
+		userAccess.Object = objectTag(target)
+		return nil
+	})
+	if err != nil {
+		return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
+	}
+	return userAccess, nil
 }
 
 // ReadUserAccessLevelForTarget returns the subject's (user) access level
 // for the given user on the given target.
-func (s *State) ReadUserAccessLevelForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
-	return corepermission.NoAccess, errors.NotImplementedf("ReadUserAccessLevelForTarget")
+func (st *State) ReadUserAccessLevelForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
+	userAccessType := corepermission.NoAccess
+	db, err := st.DB()
+	if err != nil {
+		return userAccessType, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		userUUID, err := st.userUUIDForName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Based on the grant to and grant from, find the permission,
+		// then the access type of it.
+		accessType, _, err := st.findAccessType(ctx, tx, target.Key, userUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		userAccessType = corepermission.Access(accessType)
+		return nil
+	})
+	if err != nil {
+		return userAccessType, errors.Trace(domain.CoerceError(err))
+	}
+	return userAccessType, nil
 }
 
 // ReadAllUserAccessForUser returns a slice of the user access the given
 // subject's (user) has for any access type.
-func (s *State) ReadAllUserAccessForUser(ctx context.Context, subject string) ([]corepermission.UserAccess, error) {
-	return nil, errors.NotImplementedf("ReadAllUserAccessForUser")
+func (st *State) ReadAllUserAccessForUser(ctx context.Context, subject string) ([]corepermission.UserAccess, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var (
+		permissions []readUserPermission
+		user        user
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		user, err = st.findUserByName(ctx, tx, subject)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		userPermissions, err := st.readUsersPermissions(ctx, tx, user.UUID)
```

Comment:
You should be able to write a new query for this that joins to the union, so you don't have to potato in `grantOnType`.

Reply:
@manadart Not sure what you mean by "potato" here. And I'll need help on how to combine the queries. I was happy to get what I have. :-)